test/recipes/15-test_out_option.t: add further tests

levitte opened the pull request:
Test writing to the null device.  This should be successful.
Test writing to a writable file in a write protected directory.  This
should be successful.

Also, refactor so the planned number of tests is calculated.

[extended tests]

-----

The only reason for WIP is the controversy of devnull() on some platforms or cross compilations.  I want to have a closer look at what's actually affected.

levitte committed:
test/recipes/15-test_out_option.t: add further tests

Test writing to the null device.  This should be successful.
Test writing to a writable file in a write protected directory.  This
should be successful.

Also, refactor so the planned number of tests is calculated.

[extended tests]

levitte commented:
Note that tests are currently likely to fail for `-out /dev/null` because we're checking if `/dev/` is writable.  That's flawed.

levitte commented:
This got extended to test the case of outputing to a writable file in a non-writable directory

dot-asm commented:
> Note that tests are currently likely to fail for -out /dev/null

They will also fail in true cross-compile-yet-executable-test cases. What's "true" and "false"? "True" is when current perl's devnull has no meaning to target executable. Examples are mingw targets under Linux (line on CI) and Cygwin. "False" is when current perl's devnull is applicable even in target executable context. Examples are qemu-user-assisted runs on Linux, they simply use same file system. Another example of "false" cross-compile is mingw build under MSYS2, but latter is "cheating" by transparently converting "/dev/null" to "nul".

And note the Windows "nul". It has no directory and can be viewed as omni-present in each current directory.

levitte commented:
I think I need to clarify, this change is to test things that *should* work, i.e. it's an attempt at test driven development.

> Another example of "false" cross-compile is mingw build under MSYS2, but latter is "cheating" by transparently converting "/dev/null" to "nul".

So `openssl genrsa -out /dev/null` gets silently converted, and should therefore work as long as we stop the elaborate checking.

> And note the Windows "nul". It has no directory and can be viewed as omni-present in each current directory.

So `openssl genrsa -out nul` should therefore work as long as we stop the elaborate checking.

Point is, I want this to test *as much as possible* of what is sensible under different environments.  In a MSYS2/mingw shell, `-out /dev/null` should be sensible and should therefore work.  Same thing with Cygwin.  In a Windows command prompt, `-out NUL` is sensible and should therefore work.  On VMS, `-out NLA0:` is sensible and should therefore work.

dot-asm commented:
> Same thing with Cygwin.

What same thing with Cygwin? Cygwin perl doesn't do any covert conversions, i.e. it will always pass /dev/null, be it cygwin binary or cross-compiled "mingw" one.

levitte commented:
Ok, please @dot-asm, if you want to be helpful, then come with concrete solutions.  Right now, and in all arguments we've had on this, I end up having a hopeless feeling of "no-can-do".  Is that what you're trying to tell me, that we basically shouldn't bother with the devnull check on anything Windows related?  I have a *very* *hard* *time* accepting that, so I'd like to know under what precise circumstances doing that test is possible (and hopefully meaningful under those circumstances), and under what circumstances they aren't.  If you can give me that in terms of `$^O`, `$ENV{CROSS_COMPILE}`, `config('target')`, I'd be quite grateful.

dot-asm commented:
> Ok, please @dot-asm, if you want to be helpful, then come with concrete solutions.

As for concrete solutions it would take digging into details of particular case, something I didn't do, but I figured that pointing out/reminding counter-intuitive limitations would be kind of helpful. So it doesn't help at all? In such case I'm sorry...

levitte commented:
We can start with things we commonly test:

1. build and test under windows command prompt (`VC-*`).
2. build and test under msys with mingw shell (`mingw*` in that environment) with msys perl.  As far as I've seen, there's no reason why `-out /dev/null` shouldn't work if it wasn't for the over elaborate checks `apps/openssl` does.
3. build and test under Unix (any Unix, really, but primarly Linux)
4. build and test under VMS.
5. cross-compile mingw and test the result on Linux.  *That's* an environment where I can see a reason where the devnull test is meaningless, 'cause Linux' perl has different expectations than wine...
6. testing under qemu.  This is a case I have zero experience in, so this is one where it would be helpful to detect when that *can* be an issue.
7. other ideas?

As for diverse mixtures of cygwin, msys2 and native Windows, that's way over my head for the moment, something I've never had to deal with.  Do we support that in our test recipes, or is that out of scope for them?

levitte committed:
fixup! test/recipes/15-test_out_option.t: add further tests

levitte committed:
fixup! test/recipes/15-test_out_option.t: add further tests

levitte committed:
fixup! test/recipes/15-test_out_option.t: add further tests

levitte committed:
fixup! test/recipes/15-test_out_option.t: add further tests

Reduce the amount of testing.  We only need to check that one attempt
to write to a directory fails, and we can skip checking writability to
"protected" directories.  Finally, we skip trying to check by timing,
and we can further reduce the testing code to make it look less
complex.

levitte committed:
fixup! test/recipes/15-test_out_option.t: add further tests

use "openssl rand" rather than "openssl genrsa"

levitte committed:
fixup! test/recipes/15-test_out_option.t: add further tests

dot-asm commented:
I don't quite follow, isn't this effectively replaced by #6033?

levitte commented:
Yes, it is.  The only intent with it is to show the test result before the revert, as comparison to after.  So yes, an approval of #6033 is effectively an approval of this PR, and I intend to close this when #6033 is clossed.
